ginga/util/io_fits.py

```diff
@@ -228,53 +252,39 @@ def load_hdu(self, hdu, dstobj=None, **kwargs):
         if typ == 'image':
 
             if dstobj is None:
-                # get model class for this type of object
-                obj_class = self.factory_dict.get('image', None)
-                if obj_class is None:
-                    raise FITSError(
-                        "I don't know how to load objects of kind 'image'")
-
-                dstobj = obj_class(logger=self.logger)
+                dstobj = AstroImage(logger=self.logger)
+                dstobj.load_hdu(hdu, **kwargs)
 
-            # For now, call back into the object to load it from pyfits-style
-            # HDU in future migrate to storage-neutral format
-            dstobj.load_hdu(hdu, **kwargs)
+            else:
+                # TODO: migrate code from AstroImage to here
+                dstobj.load_hdu(hdu, **kwargs)
 
         elif typ == 'table':
             # <-- data may be a table
 
-            # Handle ASDF embedded in FITS.
-            # TODO: Populate EXTNAME, EXTVER, NAXISn in ASDF meta from HDU?
-            # TODO: How to read from all the different ASDF layouts?
-            # TODO: Cache the ASDF object?
-            from ginga.util import io_asdf
-            if io_asdf.have_asdf and hdu.name == 'ASDF':
-                from asdf.fits_embed import AsdfInFits
-                from ginga import AstroImage
+            if hdu.name == 'ASDF':
+                if dstobj is not None:
+                    raise ValueError("It is not supported to load ASDF HDU with dstobj != None")
+
                 self.logger.debug('Attempting to load {} extension from '
                                   'FITS'.format(hdu.name))
-                dstobj = AstroImage.AstroImage()
-                with AsdfInFits.open(self.fits_f) as asdf_f:
-                    dstobj.load_asdf(asdf_f)
-                return dstobj
+                from ginga.util.io import io_asdf
+                opener = io_asdf.ASDFFileHandler(self.logger)
+                return opener.load_asdf_hdu_in_fits(self.fits_f, hdu, **kwargs)
 
             if dstobj is None:
-                self.logger.debug('Attempting to load table from FITS')
-                # get model class for this type of object
-                obj_class = self.factory_dict.get('table', None)
-                if obj_class is None:
-                    raise FITSError(
-                        "I don't know how to load objects of kind 'table'")
+                dstobj = AstroTable(logger=self.logger)
 
-                dstobj = obj_class(logger=self.logger)
+            self.logger.debug('Attempting to load table from FITS')
 
-            # For now, call back into the object to load it from pyfits-style
-            # HDU in future migrate to storage-neutral format
+            # TODO: migrate code from AstroTable to here
```

Comment:
Is this something you wish to address in this PR? Or is it a future PR?

Reply:
For the future.